File: app/models/tax_rate.py
import uuid
from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, ForeignKey, Numeric, String, text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.tenant import Tenant


class TaxRate(Base):
    __tablename__ = "tax_rates"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    rate: Mapped[Decimal] = mapped_column(Numeric(5, 4), nullable=False, server_default="0")
    included_in_price: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("true"))
    is_default: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("false"))
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=text("true"))

    tenant: Mapped["Tenant"] = relationship(back_populates="tax_rates")

File: php/src/Models/TaxRate.php
<?php

declare(strict_types=1);

namespace App\Models;

use App\Core\Row;

final class TaxRate
{
    /** @param string $rate fraccion decimal como string, p.ej. "0.0800" (NUMERIC(5,4)) */
    public function __construct(
        public readonly string $id,
        public readonly string $tenantId,
        public readonly string $name,
        public readonly string $rate,
        public readonly bool $includedInPrice,
        public readonly bool $isDefault,
        public readonly bool $isActive,
    ) {
    }

    /** @param array<string, mixed> $row */
    public static function fromRow(array $row): self
    {
        return new self(
            id: $row['id'],
            tenantId: $row['tenant_id'],
            name: $row['name'],
            rate: $row['rate'],
            includedInPrice: Row::bool($row['included_in_price']),
            isDefault: Row::bool($row['is_default']),
            isActive: Row::bool($row['is_active']),
        );
    }
}
